pkg/util/useragent.go

```diff
@@ -58,3 +59,25 @@ func PrefixFromUserAgent(u string) string {
 	}
 	return buf.String()
 }
+
+// UserAgentRoundTripper implements the http.RoundTripper interface and adds a User-Agent
+// header.
+type userAgentRoundTripper struct {
+	inner     http.RoundTripper
+	userAgent string
+}
+
+// UserAgentRoundTripper returns a RoundTripper that functions identically to
+// the provided 'inner' round tripper, other than also setting a user agent.
+func UserAgentRoundTripper(inner http.RoundTripper, userAgent string) http.RoundTripper {
+	return userAgentRoundTripper{
+		inner:     inner,
+		userAgent: userAgent,
+	}
+}
+
+// RoundTrip implements http.RoundTripper
+func (u userAgentRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("User-Agent", u.userAgent)
+	return u.inner.RoundTrip(req)
+}
```

Comment:
This code existed in cert-manager once before and I'm reviving it.
Here's the history:

 * Added: https://github.com/cert-manager/cert-manager/pull/422
 * Moved: https://github.com/cert-manager/cert-manager/pull/432
 * Obsoleted: https://github.com/cert-manager/cert-manager/pull/797
 * Deleted: https://github.com/cert-manager/cert-manager/pull/966

The only alteration I made was to use `req.Header.Set` rather than `req.Header.Add`,
because there should only be a single `User-Agent` header.

Other examples of this pattern:
 * https://github.com/jetstack/venafi-connection-lib/pull/124
 * https://github.com/opentofu/opentofu/blob/230fc89a28dfed5a4b0d6eed43dc8d9fc639d24b/internal/httpclient/useragent.go#L22-L34


The [RoundTripper documentation](https://pkg.go.dev/net/http#RoundTripper) says:
> // RoundTrip should not modify the request, except for
> // consuming and closing the Request's Body. RoundTrip may
> // read fields of the request in a separate goroutine. Callers
> // should not mutate or reuse the request until the Response's
> // Body has been closed.
* https://github.com/golang/go/issues/2146

This implementation does mutate the request, but I _don't think_ it will cause any problems.
And ultimately the vcert library may provide a way to set the User-Agent and their implementation will hopefully not use this RoundTripper pattern and instead add the header when they generate each request. 
Or similar to how Go ACME package implemented it:
 * https://github.com/golang/go/issues/24496
 * https://go-review.googlesource.com/c/crypto/+/183267